test/pthread/test_pthread_proxying.c

```diff
@@ -314,6 +324,186 @@ void test_proxy_callback_with_ctx(void) {
   destroy_widget(&w13);
 }
 
+em_promise_result_t explode(void** result, void* data, void* value) {
+  printf("error!");
+  abort();
+  return EM_PROMISE_REJECT;
+}
+
+em_promise_result_t test_promise_self(void** result, void* data, void* value) {
+  widget* w14 = data;
+
+  em_promise_t promise =
+    emscripten_proxy_promise(proxy_queue, pthread_self(), do_run_widget, w14);
+
+  *result = emscripten_promise_then(promise, check_widget, explode, w14);
+  emscripten_promise_destroy(promise);
+
+  return EM_PROMISE_MATCH_RELEASE;
+}
+
+em_promise_result_t
+test_promise_looper(void** result, void* data, void* value) {
+  widget* w15 = data;
+
+  em_promise_t promise =
+    emscripten_proxy_promise(proxy_queue, looper, do_run_widget, w15);
+
+  *result = emscripten_promise_then(promise, check_widget, explode, w15);
+  emscripten_promise_destroy(promise);
+
+  return EM_PROMISE_MATCH_RELEASE;
+}
+
+em_promise_result_t
+test_promise_returner(void** result, void* data, void* value) {
+  widget* w16 = data;
+
+  em_promise_t promise =
+    emscripten_proxy_promise(proxy_queue, returner, do_run_widget, w16);
+
+  *result = emscripten_promise_then(promise, check_widget, explode, w16);
+  emscripten_promise_destroy(promise);
+
+  return EM_PROMISE_MATCH_RELEASE;
+}
+
+void do_exit_thread(void* arg) { emscripten_runtime_keepalive_pop(); }
+
+em_promise_result_t exit_thread(void** result, void* data, void* value) {
+  // TODO: exit the thread directly rather than on the next turn of the event
+  // loop.
```

Comment:
Why is this a TODO?

Could we make this automatic (one day) and have the outstanding promises keep the thread alive and have the resolution of the final outstanding promise take the thread down?

Reply:
Trying to exit directly with `pthread_exit` does not work because that ends up throwing an exception to try to unwind, which in turn ends up rejecting the promise rather than gracefully exiting the thread.

Outstanding promises (at least those created with `emscripten_promise_then`) do keep the thread alive until their callbacks are executed, so this almost works. The only thing missing is that we don't use `callUserCallback` in the promise callbacks, so nothing actually checks the runtimeKeepAlive and exits. I'll fix that in a separate PR.